src/wallet/spend.cpp

```diff
@@ -320,137 +320,145 @@ CoinsResult AvailableCoins(const CWallet& wallet,
     std::vector<COutPoint> outpoints;
 
     std::set<uint256> trusted_parents;
-    for (const auto& entry : wallet.mapWallet)
-    {
-        const uint256& txid = entry.first;
-        const CWalletTx& wtx = entry.second;
+    // Cache for whether each tx passes the tx checks (first bool), and whether the transaction is "safe" (second bool)
+    std::unordered_map<uint256, std::pair<bool, bool>, SaltedTxidHasher> tx_safe_cache;
+    for (const auto& [outpoint, txo] : wallet.GetTXOs()) {
+        const CWalletTx& wtx = txo.GetWalletTx();
+        const CTxOut& output = txo.GetTxOut();
 
-        if (wallet.IsTxImmatureCoinBase(wtx) && !params.include_immature_coinbase)
+        if (tx_safe_cache.count(outpoint.hash) > 0 && !tx_safe_cache.at(outpoint.hash).first) {
             continue;
+        }
 
-        int nDepth = wallet.GetTxDepthInMainChain(wtx);
-        if (nDepth < 0)
+        // Skip manually selected coins (the caller can fetch them directly)
+        if (coinControl && coinControl->HasSelected() && coinControl->IsSelected(outpoint))
             continue;
 
-        // We should not consider coins which aren't at least in our mempool
-        // It's possible for these to be conflicted via ancestors which we may never be able to detect
-        if (nDepth == 0 && !wtx.InMempool())
+        if (wallet.IsLockedCoin(outpoint) && params.skip_locked)
             continue;
 
-        bool safeTx = CachedTxIsTrusted(wallet, wtx, trusted_parents);
-
-        // We should not consider coins from transactions that are replacing
-        // other transactions.
-        //
-        // Example: There is a transaction A which is replaced by bumpfee
-        // transaction B. In this case, we want to prevent creation of
-        // a transaction B' which spends an output of B.
-        //
-        // Reason: If transaction A were initially confirmed, transactions B
-        // and B' would no longer be valid, so the user would have to create
-        // a new transaction C to replace B'. However, in the case of a
-        // one-block reorg, transactions B' and C might BOTH be accepted,
-        // when the user only wanted one of them. Specifically, there could
-        // be a 1-block reorg away from the chain where transactions A and C
-        // were accepted to another chain where B, B', and C were all
-        // accepted.
-        if (nDepth == 0 && wtx.mapValue.count("replaces_txid")) {
-            safeTx = false;
-        }
+        if (wallet.IsSpent(outpoint))
+            continue;
```

Comment:
I have the impression that `m_txos` is also used to look up information on old transactions and we therefore want to keep all transaction outputs the wallet ever had.

Would it perhaps make sense to also have a reduced set that only keeps the potentially _unspent_ TXOs? I could see a wallet that participates in a few dozen transactions every day have magnitudes more TXOs than UTXOs. So perhaps the work the wallet has to perform every time it builds a transaction could be shortcut by storing e.g. `m_is_mature_spent` on `WalletTXO` which could be set to true when the transaction spending an output has 100 confirmations, and then either separately keeping a list of all TXOs that are not `m_is_mature_spent` or at least iterate over them more quickly directly on the basis of that member, without needing to have complex reorg handling to keep this collection up to date.

Reply:
Ah, nevermind, I just saw https://github.com/bitcoin/bitcoin/pull/27865